src/snowflake/connector/aio/_file_transfer_agent.py

```diff
@@ -0,0 +1,287 @@
+            chunk_id: int,
+        ) -> None:
+            # Note: chunk_id is 0 based while num_of_chunks is count
+            logger.debug(
+                f"Chunk {chunk_id}/{done_client.num_of_chunks} of file {done_client.meta.name} reached callback"
+            )
+            if task.exception():
+                done_client.failed_transfers += 1
+                logger.debug(
+                    f"Chunk {chunk_id} of file {done_client.meta.name} failed to transfer for unexpected exception {task.exception()}"
+                )
+            else:
+                done_client.successful_transfers += 1
+            logger.debug(
+                f"Chunk progress: {done_client.meta.name}: completed: {done_client.successful_transfers} failed: {done_client.failed_transfers} total: {done_client.num_of_chunks}"
+            )
+            if (
+                done_client.successful_transfers + done_client.failed_transfers
+                == done_client.num_of_chunks
+            ):
+                if is_upload:
+                    finish_upload_task = asyncio.create_task(
+                        done_client.finish_upload()
+                    )
+                    finish_download_upload_tasks.append(finish_upload_task)
+                    done_client.delete_client_data()
+                else:
+                    finish_download_task = asyncio.create_task(
+                        done_client.finish_download()
+                    )
+                    finish_download_task.add_done_callback(
+                        lambda t, task=finish_download_task, dc=done_client: postprocess_done_cb(
+                            task, dc
+                        )
+                    )
+                    finish_download_upload_tasks.append(finish_download_task)
+
+        def postprocess_done_cb(
+            task: asyncio.Task,
+            done_client: SnowflakeStorageClient,
+        ) -> None:
+            logger.debug(f"File {done_client.meta.name} reached postprocess callback")
+
+            if task.exception():
+                done_client.failed_transfers += 1
+                logger.debug(
+                    f"File {done_client.meta.name} failed to transfer for unexpected exception {task.exception()}"
+                )
+            # Whether there was an exception or not, we're done the file.
+
+        task_of_files = []
+        for file_client in files:
+            try:
+                res = (
+                    await file_client.prepare_upload()
+                    if is_upload
+                    else await file_client.prepare_download()
+                )
+                is_successful = True
+            except Exception as e:
+                res = e
+                is_successful = False
+
+            task = asyncio.create_task(
+                preprocess_done_cb(is_successful, res, done_client=file_client)
+            )
+            task_of_files.append(task)
+        await asyncio.gather(*task_of_files)
```

Comment:
do we have error test cases making sure that asyncio implementation in failing cases errors are raised properly?

Reply:
let's add a mock test to raise failure